drivers/soft_uart/soft_uart.c

```diff
@@ -0,0 +1,323 @@
+    ctx->data_bits = 8;
+    ctx->stop_bits = 1;
+    ctx->parity    = PARITY_NONE;
+#endif
+
+    ctx->rx_cb = rx_cb;
+    ctx->rx_cb_arg = arg;
+
+    ctx->state_rx = STATE_RX_IDLE;
+
+    if (cfg->tx_pin != GPIO_UNDEF) {
+        timer_init(cfg->tx_timer, cfg->timer_freq, _tx_timer_cb, (void*) uart);
+        gpio_write(cfg->tx_pin, !(cfg->flags & SOFT_UART_FLAG_INVERT_TX));
+        gpio_init(cfg->tx_pin, GPIO_OUT);
+    }
+
+    if (rx_cb) {
+        timer_init(cfg->rx_timer, cfg->timer_freq, _rx_timer_cb, (void*) uart);
+        timer_stop(cfg->rx_timer);
+        /* timer should fire at the end of the byte */
+        timer_set_periodic(cfg->rx_timer, 0, ctx->baud * (BITS_DATA(ctx) + BITS_PARITY(ctx) + 1),
+                           TIM_FLAG_RESET_ON_MATCH | TIM_FLAG_RESET_ON_SET);
+        gpio_init_int(cfg->rx_pin, GPIO_IN, GPIO_BOTH, _rx_gpio_cb, (void*) uart);
+    }
+
+    return 0;
+}
+
+#ifdef MODULE_SOFT_UART_MODECFG
+int soft_uart_mode(soft_uart_t uart, uart_data_bits_t data_bits, uart_parity_t parity,
+                   uart_stop_bits_t stop_bits)
+{
+    if (uart >= SOFT_UART_NUMOF) {
+        return UART_NODEV;
+    }
+
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    switch (data_bits) {
+    case UART_DATA_BITS_5:
+        ctx->data_bits = 5;
+        break;
+    case UART_DATA_BITS_6:
+        ctx->data_bits = 6;
+        break;
+    case UART_DATA_BITS_7:
+        ctx->data_bits = 7;
+        break;
+    case UART_DATA_BITS_8:
+        ctx->data_bits = 8;
+        break;
+    }
+
+    switch (parity) {
+    case UART_PARITY_NONE:
+        ctx->parity = PARITY_NONE;
+        break;
+    case UART_PARITY_EVEN:
+        ctx->parity = PARITY_EVEN;
+        break;
+    case UART_PARITY_ODD:
+        ctx->parity = PARITY_ODD;
+        break;
+    case UART_PARITY_MARK:
+        ctx->parity = PARITY_MARK;
+        break;
+    case UART_PARITY_SPACE:
+        ctx->parity = PARITY_SPACE;
+        break;
+    }
+
+    switch (stop_bits) {
+    case UART_STOP_BITS_1:
+        ctx->stop_bits = 1;
+        break;
+    case UART_STOP_BITS_2:
+        ctx->stop_bits = 2;
+        break;
+    }
+
+    return 0;
+}
+#endif /* MODULE_SOFT_UART_MODECF */
+
+static void soft_uart_write_byte(soft_uart_t uart, uint8_t data)
+{
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    /* start bit (LOW) + data bits */
+    ctx->bits_tx = 1 + BITS_DATA(ctx);
+    ctx->byte_tx = data << 1;
+
+#ifdef MODULE_SOFT_UART_MODECFG
+    if (ctx->parity != PARITY_NONE) {
+        uint8_t parity = 0;
+
+        switch (ctx->parity) {
+        case PARITY_EVEN:
+            parity = __builtin_parity(data);
+            break;
+        case PARITY_ODD:
+            parity = !__builtin_parity(data);
+            break;
+        case PARITY_MARK:
+            parity = 1;
+            break;
+        case PARITY_SPACE:
+            parity = 0;
+            break;
+        }
```

Comment:
I personally like the indent of this. But the coding convention not :-/

Reply:
Huh? 

> - Please obey the Linux coding style as described in https://www.kernel.org/doc/Documentation/process/coding-style.rst with the following modifications and additions:
> (nothing about indentation in switch statements)


#### coding-style.rst

```
The preferred way to ease multiple indentation levels in a switch statement is
to align the ``switch`` and its subordinate ``case`` labels in the same column
instead of ``double-indenting`` the ``case`` labels.  E.g.:

```C
	switch (suffix) {
	case 'G':
	case 'g':
		mem <<= 30;
		break;
	case 'M':
	case 'm':
		mem <<= 20;
		break;
	case 'K':
	case 'k':
		mem <<= 10;
		fallthrough;
	default:
		break;
	}
```
```

Reply:
Uncrustify enforces this to be different. I assumed this to be reflected by the coding convention.

<details><summary>Test with uncrustify</summary>

```
$ cat test.c
#include <stddef.h>

size_t foo(char suffix)
{
    size_t mem = 1;
    switch (suffix) {
    case 'G':
    case 'g':
        mem <<= 30;
        break;
    case 'M':
    case 'm':
        mem <<= 20;
        break;
    case 'K':
    case 'k':
        mem <<= 10;
        fallthrough;
    default:
        break;
    }

    return mem;
}

$ uncrustify -c uncrustify-riot.cfg test.c 
Output suffix: .uncrustify
Parsing: test.c as language C

$ diff -p test.c test.c.uncrustify 
--- test.c
+++ test.c.uncrustify
@@ -3,21 +3,22 @@
 size_t foo(char suffix)
 {
     size_t mem = 1;
+
     switch (suffix) {
-    case 'G':
-    case 'g':
-        mem <<= 30;
-        break;
-    case 'M':
-    case 'm':
-        mem <<= 20;
-        break;
-    case 'K':
-    case 'k':
-        mem <<= 10;
-        fallthrough;
-    default:
-        break;
+        case 'G':
+        case 'g':
+            mem <<= 30;
+            break;
+        case 'M':
+        case 'm':
+            mem <<= 20;
+            break;
+        case 'K':
+        case 'k':
+            mem <<= 10;
+            fallthrough;
+        default:
+            break;
     }
 
     return mem;
```

</details>

Reply:
If you can get this past Murdock, I'm all in for it ;-) As you pointed out, this is indeed exactly as the coding convention says.

Reply:
I would assume this to be a bug in the uncrustify configuration :D

Reply:
IIRC uncrustify is only enforced for `core/` - did this change after the feature freeze?

Reply:
If it is not enforced, this should be fine :-)

Reply:
Travis says

> 	All files are uncrustified!